lib/json-schema/validator.rb

```diff
@@ -238,6 +242,46 @@ def validation_errors
       @errors
     end
 
+    def resolve_nested_references(schema, schema_portion, list)
+      case schema_portion
+      when Array
+        # for arrays resolve all items which have not been resolved yet
+        schema_portion.each_with_index do |item, i|
+          schema_portion[i] = resolve_nested_references(schema, item, list) unless list.include?(schema_portion[i])
+        end
+      when Hash
+        if schema_portion["$ref"]
+          # for hashes with a reference, resolve their referenced schema and copy(!) the result into the currently handled portion
+          _, ref_schema = JSON::Schema::RefAttribute.get_referenced_uri_and_schema(schema_portion, schema, self)
+          resolved_portion = resolve(ref_schema, ref_schema.schema, list)
+          schema_portion.clear
+          resolved_portion.each { |key, value| schema_portion[key] = value }
+        else
+          # for hashes without a reference we also resolve recursively what has not been resolved yet
+          schema_portion.each do |key, value|
+            schema_portion[key] = resolve_nested_references(schema, value, list) unless list.include?(value)
+          end
+        end
+      end
+      schema_portion
+    end
+
+    def resolve(schema, handle_schema_hash, list = [])
```

Comment:
It's not clear to me what `handle_schema_hash` is, and again, I think that `list` could have a better name.